tools/codegen/src/generator/languages/rust_lazy_reader/generator.rs

```diff
@@ -0,0 +1,718 @@
+                            if cur.option_is_none() {
+                                Ok(None)
+                            } else {
+                                let cur = cur.convert_to_rawbytes()?;
+                                Ok(Some(cur.into()))
+                            }
+                        }
+                    } else if has_from {
+                        quote! {
+                            if cur.option_is_none() {
+                                Ok(None)
+                            } else {
+                                Ok(Some(cur.into()))
+                            }
+                        }
+                    } else {
+                        quote! {
+                            if cur.option_is_none() {
+                                Ok(None)
+                            } else {
+                                Ok(Some(cur.try_into()?))
+                            }
+                        }
+                    }
+                } else if level == 2 {
+                    if flag {
+                        quote! {
+                            if cur.option_is_none() {
+                                Ok(None)
+                            } else {
+                                let cur = cur.convert_to_rawbytes()?;
+                                Ok(Some(Some(cur.try_into())))
+                           }
+                        }
+                    } else {
+                        quote! {
+                            if cur.option_is_none() {
+                                Ok(None)
+                            } else {
+                                Ok(Some(Some(cur.into())))
+                            }
+                        }
+                    }
+                } else {
+                    panic!("can't support")
+                }
+            }
+            TypeCategory::Type => quote! { Ok(cur.into()) },
+            TypeCategory::Primitive => quote! { cur.try_into() },
+            TypeCategory::Array => quote! { Ok(cur) },
+            TypeCategory::FixVec => {
+                quote! {
+                    cur.convert_to_rawbytes()
+                }
+            }
+        }
+    }
+}
+
+// see TypeCategory
+fn get_rust_type_category(typ: &TopDecl) -> (TokenStream, TypeCategory) {
+    let name = typ.name();
+    let mut tc = TypeCategory::Primitive;
+    let token_name = ident_new(name);
+    let mut transformed_name = quote! { #token_name };
+    match typ {
+        // if the field type is array and the field type name is "uint8", "int8" ...
+        // then it's primitive
+        TopDecl::Array(a) => {
+            let field_type_name = name.to_lowercase();
+            let new_name = match field_type_name.as_ref() {
+                // see https://github.com/XuJiandong/moleculec-c2#extra-support-for-known-types
+                "uint8" => quote! { u8 },
+                "int8" => quote! { i8 },
+                "uint16" => quote! { u16 },
```

Comment:
I would personally suggest here we generate 3 methods for each field like `uint16`:

* A method returns `Uint16` structure directly
* A second method returning `u16` decoding it in little endian encoding
* A third method returning `u16` decoding it in big endian encoding

Reply:
Also, we were generating handy shortcut when a table includes a `Uint16` type, but for `Uint16` type itself, we were not generating any shortcut for getting a `u16`, it might be better to have shortcut methods in both places.